cmd/metal-api/internal/headscale/client.go

```diff
@@ -0,0 +1,238 @@
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	timestamppb "google.golang.org/protobuf/types/known/timestamppb"
+)
+
+var (
+	apiKeyExpiration = 24 * 90 * time.Hour
+)
+
+type HeadscaleClient struct {
+	client headscalev1.HeadscaleServiceClient
+
+	address             string
+	ControlPlaneAddress string
+
+	apiKeyPrefix    string
+	oldAPIKeyPrefix string
+
+	mu         sync.Mutex
+	wg         sync.WaitGroup
+	ctx        context.Context
+	conn       *grpc.ClientConn
+	cancelFunc context.CancelFunc
+	logger     *zap.SugaredLogger
+}
+
+func NewHeadscaleClient(addr, controlPlaneAddr, apiKey string, logger *zap.SugaredLogger) (client *HeadscaleClient, err error) {
+	if addr != "" || apiKey != "" {
+		if addr == "" {
+			return nil, fmt.Errorf("headscale address should be set with api key")
+		}
+		if apiKey == "" {
+			return nil, fmt.Errorf("headscale api key should be set with address")
+		}
+	} else {
+		return
+	}
+
+	h := &HeadscaleClient{
+		address:             addr,
+		ControlPlaneAddress: controlPlaneAddr,
+
+		mu:     sync.Mutex{},
+		wg:     sync.WaitGroup{},
+		logger: logger,
+	}
+	h.ctx, h.cancelFunc = context.WithCancel(context.Background())
+
+	// Validate API Key
+	if h.apiKeyPrefix, err = getApiKeyPrefix(apiKey); err != nil {
+		return nil, err
+	}
+
+	if err = h.connect(apiKey); err != nil {
+		return nil, fmt.Errorf("failed to connect to Headscale server: %w", err)
+	}
+
+	if err = h.replaceApiKey(); err != nil {
+		return nil, fmt.Errorf("failed to replace Headscale API Key: %w", err)
+	}
+
+	go h.rotateApiKey()
+
+	return h, nil
+}
+
+// Connect or reconnect to Headscale server
+func (h *HeadscaleClient) connect(apiKey string) (err error) {
+	ctx, cancel := context.WithTimeout(h.ctx, 5*time.Second)
+	defer cancel()
+
+	grpcOptions := []grpc.DialOption{
+		grpc.WithBlock(),
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithPerRPCCredentials(tokenAuth{
+			token: apiKey,
+		}),
+	}
+
+	h.conn, err = grpc.DialContext(ctx, h.address, grpcOptions...)
+	if err != nil {
+		return fmt.Errorf("failed to connect to headscale server %s: %w", h.address, err)
+	}
+
+	h.client = headscalev1.NewHeadscaleServiceClient(h.conn)
+
+	return
+}
+
+// Refresh API key when apiKeyExpiration reaches it's half life
+func (h *HeadscaleClient) rotateApiKey() {
+	h.wg.Add(1)
+	defer h.wg.Done()
+
+	ticker := time.NewTicker(apiKeyExpiration / 2)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-h.ctx.Done():
+			return
+		case <-ticker.C:
+			// Try replacement until it works
+			// Log error messages
+			for {
+				if err := h.replaceApiKey(); err != nil {
+					h.logger.Error("failed to replace Headscale API Key", zap.Error(err))
+					time.Sleep(time.Second)
+					continue
+				}
+
+				break
+			}
+		}
+	}
+}
+
+// Expires current API Key and replaces it with a new one
+func (h *HeadscaleClient) replaceApiKey() (err error) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
```

Comment:
Probably it's fine to remove these locks from the functions, it's really unlikely that this will cause problems.